### DISCUSSION TITLE
Trivy does not acknowledge custom policy

### DISCUSSION THREAD

**Discussion by l3112** [aquasecurity/trivy#4464] [Q&A] [answered]
### Question

I have a custom policy; Trivy doesn't see it, so it doesn't apply it to my Terraform file.
- When it was in the same file, I would call the policy; It would not read it.
-  When I dug around in Library/caches/Trivy and put it under /Library/Caches/trivy/policy/content/policies/cloud/policies/aws/s3, it would not read it.

Trivy does run the default misconfig checks, so it works.

Where do I put this policy so Trivy sees it?

### Target

A Terraform main.tf file I see this info [over here ](https://github.com/aquasecurity/trivy/pull/4457/files), but I'm not sure how to implement it -- Does Trivy not scan root files  with new policies by default?

### Scanner

Misconfiguration

### Output Format

None

### Mode

Standalone

### Operating System

masOS Ventura 13.3.1

### Version

```bash
0.41.0
```

**Comment by AnaisUrlichs** [comment]
Hi there, I think you are trying to do multiple different things 

Generally, if you write your custom rego policy, you have to tell trivy where that policy is that you want to add to the misconfiguration scan such as the Terraform scan. This is done with the `--policy` flag:
`trivy config --policy ./path-to-policy --namespace <your namespace> ./path-to-resource-you-want-to-scan`

Trivy will then apply the rego policy.  Here is the docs on custom policies https://aquasecurity.github.io/trivy/v0.41/docs/scanner/misconfiguration/custom/ Please let us know if something is unclear.

> When I dug around in Library/caches/Trivy and put it under /Library/Caches/trivy/policy/content/policies/cloud/policies/aws/s3, it would not read it.

Generally, you would not modify the Trivy-specific files directly but configure the Trivy CLI through its flags

> -- Does Trivy not scan root files with new policies by default?

You have to specify any custom policies for Trivy. The PR that you linked is related to a current bug we have with Terraform scanning whereby Trivy does not scan subdirectories properly

**Reply by l3112** [reply]
Let me outline the commands I'm taking;

- I'm in the same folder as main.tf in the terminal
- policies are in a policies folder in the same folder

`trivy config --policy ./policies/ main.tf` 
`trivy config --policy ./policies --namespaces user  main.tf`

It does run, and it doesn't return a deny message or point out the error that the documentation says it should and that I've written into the [policy here](https://github.com/l3112/Rego-Trivy/blob/main/ID002.rego).:

`The result of a call to result.new(msg, cause). The msg is a string describing the issue occurrence, and the cause is the property/object where the issue occurred. Providing this allows Trivy to ascertain line numbers and highlight code in the output.`

Do you have an idea as to why that is?

If the [rego (here)](https://github.com/l3112/Rego-Trivy/blob/main/ID002.rego) is faulty, does it ignore the policy?

**Comment by simar7** [comment, marked as answer]
Please see my previous reply here https://github.com/aquasecurity/trivy/discussions/4254#discussioncomment-5878829

You could use the OPA playground to debug your custom policy. My hunch is that your policy isn't getting applied because it might not be correct.
